### ISSUE TITLE
Java Full GC causing extreme lags.

### ISSUE THREAD

**Issue by Mirviy** [PaperMC/Paper#2918]
###  Timings or Profile link:
Timings may be not relevant to this issue. But I also did one:
[timings](https://timings.aikar.co/?id=d1e8d787294b4518b8058a159e6ee413)

### Description of issue:
After server started, the OldGen of java heap memory increases linearly with time, then reaches the limit (about 3.5G, in 4 hours) and then do Full GC continuously(repeatedly). Every Full GC stops the server for more than 10 seconds, in this long time server is not be able to finish 1 single tick ( tps<0.1 ). 
This may be a memory leak.

### Plugin list:
ProtocolLib 4.5.0
SuperVanish 6.1.3
MySkin 3.0
AuthMe 5.6.0-SNAPSHOT-b2352
CoreProtect 2.17.5
ColorMOTD 1.3
OpenInv 4.0.8
GroupManager 0.3.0
FastLogin 1.11-SNAPSHOT-2c7e569

### bukkit.yml, spigot.yml, paper.yml, server.properties
[bukkit](https://pastebin.com/zUUYCiXr)
[spigot](https://pastebin.com/ZGE3iJdZ)
[paper](https://pastebin.com/gBN5pmFr)
[server](https://pastebin.com/me5Tssqs)

### Other helpful links
java gc log with "
 -verbose:gc
 -XX:+PrintGCDetails
 -XX:+PrintGCTimeStamps
 -XX:+PrintGCDateStamps
 -Xloggc:gc.log
 -XX:+UseG1GC":
[gclog](https://www.kamine.cloud/files/gc.7z)

### Paper build number:
git-Paper-84 (MC: 1.15.2) (Implementing API version 1.15.2-R0.1-SNAPSHOT)

**Comment by electronicboy** [comment]
gc logs confirm that you have high base memory usage, you'll need to look into analyzing heap dumps to work out what is actually using memory

**Comment by themetha** [comment]
Also noticed a similar issue where server runs out of memory and just runs full gc.
Did a heap dump and looked through it.
It seems that there is an abnormal amount of objects in net.minecraft.server.v1_15_R1.Entity[] . In the specific one I checked I had 2^15 endermen, all dead.
It seems this list is not unreferenced and garbage collected properly.

In my case, the structure looked like:

```

Class Name                                                         | Shallow Heap | Retained Heap | Percentage
---------------------------------------------------------------------------------------------------------------
net.minecraft.server.v1_15_R1.Chunk @ 0x6d8861f68                  |          136 |   251 831 128 |     9,17 %
|- com.destroystokyo.paper.util.maplist.EntityList @ 0x6d886cef8   |           24 |   251 792 760 |     9,17 %
|  |- net.minecraft.server.v1_15_R1.Entity[32768] @ 0x72506ff10    |      131 088 |   251 268 336 |     9,15 %
|  |  |- net.minecraft.server.v1_15_R1.EntityEnderman @ 0x75ceb7860|          824 |        64 096 |     0,00 %
---------------------------------------------------------------------------------------------------------------
```

I can supply heap dump if needed.
Also, running just paper, no additional plugins.

**Comment by electronicboy** [comment]
if you go to the chunks where said entities are supposed to be, do they all disappear? Not really sure how that would happen unless the chunks weren't ticking properly for some reason

**Comment by themetha** [comment]
Haven't tried that. Since the issue was hanging the server, I had to restart it. Will try if it happens again.
In my dump, there are multiple chunks which seems to have old entities still in memory.
I think the example mentioned above is an enderman farm in the end. I don't see why it would have 32k dead endermen if no one is in the chunk. Then it should either not spawn any new (at least not that many) or they would have expired when someone visited to the farm. Chunk unloading?

**Comment by themetha** [comment]
Updated to build 86 now to see if that is the issue. Will get back if reproduced again.

**Comment by MrPowerGamerBR** [comment]
I have the same issue, I made a server heap dump and looks like Paper isn't clearing up the NMS `entities` list, one of my chunks in the heap dump had 2GBs of retained memory and 160k entities! Teleporting to the chunk after the restart didn't show any entities, so looks like they aren't written to disk.

I *do* have a plugin that does stuff on the affected chunks: They automatically spawn entities, if the entities goes outside of the region, the entity is removed (via `remove()`) and respawned on the right place, maybe Paper isn't cleaning up mobs that were removed via the `remove()` method? Here's the plugin source code: https://github.com/PerfectDreams/DreamMobSpawner/blob/master/src/main/kotlin/net/perfectdreams/dreammobspawner/DreamMobSpawner.kt#L39

I can also share the heap dump if needed (12GBs heap dump 😭)

I also updated to build 86 and will see if the issue persists or not.
